index.js

```diff
@@ -27,6 +27,9 @@ const plugins = [
   'react',
 ];
 
+const SEVERITY_ERROR = /** @type {2} */ (2);
+const SEVERITY_OFF = /** @type {0} */ (0);
```

Comment:
can you elaborate on why this is needed/preferred, vs a literal number?

Reply:
In `eslint`'s type define ([link](https://github.com/eslint/eslint/blob/7dd402d33226d821a17b22c4753ce9c40fc909bd/lib/types/index.d.ts#L898)), a severity number can only be literal `0`, `1` or `2`. 

If we have `'react/display-name': number;` in `eslint-plugin-react`'s `index.d.ts`, we will get the error below:

```
Property ''react/display-name'' is incompatible with index signature.
  Type 'number' is not assignable to type 'RuleEntry<any[]> | undefined'.ts(2322) 
```

(This error was mentioned in [this comment](https://github.com/jsx-eslint/eslint-plugin-react/issues/3838#issuecomment-2386970537))

Since `RuleEntry` is `0 | 1 | 2 | "off" | "warn" | "error"` or an array. 

Once we have a literal number `'react/display-name': 2;` than TypeScript won't complent anymore.

Reply:
I guess I’m confused why TS infers number here when it’s a literal 2, but that makes sense.

Reply:
> why TS infers number here when it’s a literal 2

If TS infers 2 as a literal 2, the following code would break:

```
const obj = { a: 1, b: 2 }
obj.b = 3 // Cannot assign 3 to 2
```

TS provides a handy tool `as const` to address such issue, sadly we don't have similar thing in JSDoc.

```ts
const obj = { a: 1, b: 2 }
     // ^: { a: number; b: number }
const obj2 = { a: 1, b: 2 } as const 
     // ^: { a: readonly 1; b: readonly 2 }
```

Reply:
> TS provides a handy tool `as const` to address such issue, sadly we don't have similar thing in JSDoc.

We actually do, since [TypeScript v4.5](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-5.html#const-assertions-and-default-type-arguments-in-jsdoc):

https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#casts

Reply:
Thank you for the correction! I tried that but it didn't work. I realized that's because this project is still using TS v3.  Since the current approach works I'd prefer not updating TS in this PR.